docs/The State Mechanism.ipynb

```diff
@@ -0,0 +1,618 @@
+    "from autora.variable import VariableCollection, Variable"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## `State` objects\n",
+    "\n",
+    "`State` objects contain metadata describing an experiment, and the data gathered during an experiment. Any `State` \n",
+    "object used in an AutoRA cycle will be a subclass of the `autora.state.State`, with the necessary fields specified. \n",
+    "(The `autora.state.StandardState` provides some sensible defaults.)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@dataclass(frozen=True)\n",
+    "class BasicState(autora.state.State):\n",
+    "   data: pd.DataFrame = field(default_factory=pd.DataFrame, metadata={\"delta\": \"extend\"})\n",
+    "   \n",
+    "s = BasicState()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Because it is a python dataclass, the `State` fields can be accessed using attribute notation, for example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
+       "<div>\n",
+       "<style scoped>\n",
+       "    .dataframe tbody tr th:only-of-type {\n",
+       "        vertical-align: middle;\n",
+       "    }\n",
+       "\n",
+       "    .dataframe tbody tr th {\n",
+       "        vertical-align: top;\n",
+       "    }\n",
+       "\n",
+       "    .dataframe thead th {\n",
+       "        text-align: right;\n",
+       "    }\n",
+       "</style>\n",
+       "<table border=\"1\" class=\"dataframe\">\n",
+       "  <thead>\n",
+       "    <tr style=\"text-align: right;\">\n",
+       "      <th></th>\n",
+       "    </tr>\n",
+       "  </thead>\n",
+       "  <tbody>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "Empty DataFrame\n",
+       "Columns: []\n",
+       "Index: []"
+      ]
+     },
+     "execution_count": null,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "s.data  # an empty DataFrame with a column \"x\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`State` objects can be updated by adding `Delta` objects. A `Delta` represents new data, and is combined with the \n",
+    "existing data in the `State` object. The `State` itself is immutable by design, so adding a `Delta` to it creates a new \n",
+    "`State`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "BasicState(data=   x  y\n",
+       "0  1  1)"
+      ]
+     },
+     "execution_count": null,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "s + autora.state.Delta(data=pd.DataFrame({\"x\":[1], \"y\":[1]}))"
```

Comment:
is this supposed to be s_0 = ... or no?

Reply:
No, the text which follows it is wrong. I've added an update. Does this make more sense?

Reply:
yes, it does!